File: src/main/java/me/miki/shindo/ui/comp/impl/CompColorPicker.java
package me.miki.shindo.ui.comp.impl;

import lombok.Getter;
import lombok.Setter;
import me.miki.shindo.management.settings.impl.ColorSetting;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.ui.comp.Comp;
import me.miki.shindo.ui.framework.UIContext;
import me.miki.shindo.utils.MathUtils;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import me.miki.shindo.utils.mouse.MouseUtils;
import net.minecraft.util.ResourceLocation;

import java.awt.*;

public class CompColorPicker extends Comp {

    private final SimpleAnimation openAnimation = new SimpleAnimation();
    private final SimpleAnimation hueAnimation = new SimpleAnimation();
    private final SimpleAnimation saturationAnimation = new SimpleAnimation();
    private final SimpleAnimation brightnessAnimation = new SimpleAnimation();
    private final SimpleAnimation alphaAnimation = new SimpleAnimation();

    private final ColorSetting colorSetting;
    @Setter
    @Getter
    private boolean open;

    @Setter
    @Getter
    private float scale;

    private boolean hueDragging, sbDragging, alphaDragging;

    public CompColorPicker(float x, float y, ColorSetting setting) {
        super(x, y);
        this.colorSetting = setting;
        this.scale = 1.0F;
        this.open = false;
    }

    public CompColorPicker(ColorSetting setting) {
        super(0, 0);
        this.colorSetting = setting;
        this.scale = 1.0F;
        this.open = false;
    }

    @Override
    public void draw(int mouseX, int mouseY, float partialTicks) {

        NanoVGManager nvg = ctx().nvg();

        openAnimation.setAnimation(open ? 1.1F : 0.0F, 16);

        nvg.save();
        nvg.translate(0, 26 * scale);

        float hueMaxValue = 1.0F;
        float hueMinValue = 0.0F;
        float hueValue = colorSetting.getHue();
        float size = 100 * scale;

        nvg.scale(this.getX(), this.getY(), size, 0, Math.max(0, openAnimation.getValue() - 0.1F));

        float hueValueHeight = (size - (12 * scale)) * (hueValue - hueMinValue) / (hueMaxValue - hueMinValue);

        double hueDiff = Math.min(size, Math.max(0, mouseY - (this.getY() + (28 * scale))));

        hueAnimation.setAnimation(hueValueHeight, 16);

        if (hueDragging) {
            if (hueDiff == 0) {
                colorSetting.setHue(hueMinValue);
            } else {
                colorSetting.setHue((float) MathUtils.roundToPlace(((hueDiff / size) * (hueMaxValue - hueMinValue) + hueMinValue), 2));
            }
        }

        float sbMaxValue = 1.0F;
        float sbMinValue = 0.0F;
        float saturationValue = colorSetting.getSaturation();
        float brightnessValue = colorSetting.getBrightness();

        float saturationValueSize = (size - (12 * scale)) * (saturationValue - sbMinValue) / (sbMaxValue - sbMinValue);
        float brightnessValueSize = (size - (12 * scale)) * (brightnessValue - sbMinValue) / (sbMaxValue - sbMinValue);

        double brightnessDiff = Math.min(size, Math.max(0, (this.getY() + (30 * scale)) + size - mouseY));
        double saturationDiff = Math.min(size, Math.max(0, mouseX - (this.getX() - (4 * scale))));

        brightnessAnimation.setAnimation(brightnessValueSize, 20);
        saturationAnimation.setAnimation(saturationValueSize, 20);

        if (sbDragging) {

            if (brightnessDiff == 0) {
                colorSetting.setBrightness(sbMinValue);
            } else {
                colorSetting.setBrightness((float) MathUtils.roundToPlace(((brightnessDiff / size) * (sbMaxValue - sbMinValue) + sbMinValue), 2));
            }

            if (saturationDiff == 0) {
                colorSetting.setSaturation(sbMinValue);
            } else {
                colorSetting.setSaturation((float) MathUtils.roundToPlace(((saturationDiff / size) * (sbMaxValue - sbMinValue) + sbMinValue), 2));
            }
        }

        float alphaMaxValue = 255;
        float alphaMinValue = 0;
        float alphaValue = colorSetting.getAlpha();
        float alphaWidth = size + (18 * scale);

        double alphaDiff = Math.min(alphaWidth, Math.max(0, mouseX - (this.getX() - (4 * scale))));

        float alphaValueSize = (alphaWidth - (12 * scale)) * (alphaValue - alphaMinValue) / (alphaMaxValue - alphaMinValue);

        alphaAnimation.setAnimation(alphaValueSize, 20);

        if (alphaDragging) {

            if (colorSetting.isShowAlpha()) {
                if (alphaDiff == 0) {
                    colorSetting.setAlpha(0);
                } else {
                    colorSetting.setAlpha((int) MathUtils.roundToPlace(((alphaDiff / alphaWidth) * (alphaMaxValue - alphaMinValue) + alphaMinValue), 2));
                }
            } else {
                colorSetting.setAlpha(255);
            }
        }

        nvg.drawHSBBox(this.getX(), this.getY(), size, size, 6F * scale, Color.getHSBColor(colorSetting.getHue(), 1, 1));
        nvg.drawRoundedImage(new ResourceLocation("shindo/hue.png"), this.getX() + (106 * scale), this.getY(), 12 * scale, size, 3 * scale);
        nvg.drawArc(this.getX() + (112 * scale), this.getY() + hueAnimation.getValue() + (6 * scale), 3 * scale, 0, 360, 1.2F * scale, Color.WHITE);
        nvg.drawArc(this.getX() + saturationAnimation.getValue() + (6 * scale), this.getY() + size - brightnessAnimation.getValue() - (6 * scale), 3 * scale, 0, 360, 1.2F * scale, Color.WHITE);

        if (colorSetting.isShowAlpha()) {
            nvg.drawRoundedImage(new ResourceLocation("shindo/alpha.png"), this.getX(), this.getY() + (106 * scale), size + (18 * scale), 12 * scale, 3 * scale);
            nvg.drawAlphaBar(this.getX(), this.getY() + (106 * scale), alphaWidth, 12 * scale, 3 * scale, Color.getHSBColor(colorSetting.getHue(), 1, 1));
            nvg.drawArc(this.getX() + alphaAnimation.getValue() + (6 * scale), this.getY() + (112 * scale), 3 * scale, 0, 360, 1.2F * scale, Color.WHITE);
        }

        nvg.restore();

        nvg.drawRoundedRect(this.getX() + (106 * scale), this.getY(), 16 * scale, 16 * scale, 4, colorSetting.getColor());

        super.draw(mouseX, mouseY, partialTicks);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {

        if (open) {
            float size = 100 * scale;
            float alphaWidth = size + (18 * scale);
            float addY = 26 * scale;

            if (mouseButton == 0) {

                if (MouseUtils.isInside(mouseX, mouseY, this.getX() + (106 * scale), this.getY() + addY, 12 * scale, size)) {
                    hueDragging = true;
                }

                if (MouseUtils.isInside(mouseX, mouseY, this.getX(), this.getY() + addY, size, size)) {
                    sbDragging = true;
                }

                if (MouseUtils.isInside(mouseX, mouseY, this.getX(), this.getY() + (106 * scale) + addY, alphaWidth, 12 * scale) && colorSetting.isShowAlpha()) {
                    alphaDragging = true;
                }
            }
        }

        if (mouseButton == 0) {

            if (isInsideOpen(mouseX, mouseY)) {
                open = !open;
            }
        }

        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
        hueDragging = false;
        sbDragging = false;
        alphaDragging = false;
        super.mouseReleased(mouseX, mouseY, mouseButton);
    }

    public boolean isShowAlpha() {
        return colorSetting.isShowAlpha();
    }

    public boolean isInsideOpen(int mouseX, int mouseY) {
        return MouseUtils.isInside(mouseX, mouseY, this.getX() + (106 * scale), this.getY(), 16 * scale, 16 * scale);
    }
}

File: src/main/kotlin/me/miki/shindo/ui/components/v2/inputs/CompColorPicker.kt
package me.miki.shindo.ui.components.v2.inputs

import me.miki.shindo.management.settings.impl.ColorSetting
import me.miki.shindo.ui.animation.v2.value.SimpleAnimation
import me.miki.shindo.ui.components.v2.Component
import me.miki.shindo.utils.MathUtils
import me.miki.shindo.utils.mouse.MouseUtils
import net.minecraft.util.ResourceLocation
import java.awt.Color
import kotlin.math.max
import kotlin.math.min

class CompColorPicker : Component {
    private val openAnimation = SimpleAnimation()
    private val hueAnimation = SimpleAnimation()
    private val saturationAnimation = SimpleAnimation()
    private val brightnessAnimation = SimpleAnimation()
    private val alphaAnimation = SimpleAnimation()

    private val colorSetting: ColorSetting
    private var isOpen: Boolean = false
    private var scale: Float = 1.0f

    private var hueDragging = false
    private var sbDragging = false
    private var alphaDragging = false

    constructor(x: Float, y: Float, setting: ColorSetting) : super(x, y) {
        this.colorSetting = setting
        scale = 1.0f
        this.isOpen = false
    }

    constructor(setting: ColorSetting) : super(0f, 0f) {
        this.colorSetting = setting
        scale = 1.0f
        this.isOpen = false
    }

    fun isOpen(): Boolean = isOpen
    fun getScale(): Float = scale

    fun setOpen(isOpen: Boolean) {
        this.isOpen = isOpen
    }

    fun setScale(scale: Float) {
        this.scale = scale
    }

    override fun draw(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val nvgInstance = nvg

        openAnimation.setAnimation(if (isOpen) 1.1f else 0.0f, 16.0)

        nvgInstance.save()
        nvgInstance.translate(0f, 26 * scale)

        val hueMaxValue = 1.0f
        val hueMinValue = 0.0f
        val hueValue = colorSetting.getHue()
        val size = 100 * scale

        nvgInstance.scale(getX(), getY(), size, 0f, max(0f, openAnimation.getValue() - 0.1f))

        val hueValueHeight = (size - 12 * scale) * (hueValue - hueMinValue) / (hueMaxValue - hueMinValue)
        val hueDiff = min(size.toDouble(), max(0.0, mouseY - (getY() + 28 * scale).toDouble()))

        hueAnimation.setAnimation(hueValueHeight, 16.0)

        if (hueDragging) {
            if (hueDiff == 0.0) {
                colorSetting.setHue(hueMinValue)
            } else {
                colorSetting.setHue(
                    MathUtils.roundToPlace(
                        (hueDiff / size) * (hueMaxValue - hueMinValue) + hueMinValue,
                        2
                    ).toFloat()
                )
            }
        }

        val sbMaxValue = 1.0f
        val sbMinValue = 0.0f
        val saturationValue = colorSetting.getSaturation()
        val brightnessValue = colorSetting.getBrightness()

        val saturationValueSize = (size - 12 * scale) * (saturationValue - sbMinValue) / (sbMaxValue - sbMinValue)
        val brightnessValueSize = (size - 12 * scale) * (brightnessValue - sbMinValue) / (sbMaxValue - sbMinValue)

        val brightnessDiff = min(size.toDouble(), max(0.0, (getY() + 30 * scale + size - mouseY).toDouble()))
        val saturationDiff = min(size.toDouble(), max(0.0, mouseX - (getX() - 4 * scale).toDouble()))

        brightnessAnimation.setAnimation(brightnessValueSize, 20.0)
        saturationAnimation.setAnimation(saturationValueSize, 20.0)

        if (sbDragging) {
            colorSetting.setBrightness(
                if (brightnessDiff == 0.0) sbMinValue
                else MathUtils.roundToPlace((brightnessDiff / size) * (sbMaxValue - sbMinValue) + sbMinValue, 2)
                    .toFloat()
            )

            colorSetting.setSaturation(
                if (saturationDiff == 0.0) sbMinValue
                else MathUtils.roundToPlace((saturationDiff / size) * (sbMaxValue - sbMinValue) + sbMinValue, 2)
                    .toFloat()
            )
        }

        val alphaMaxValue = 255
        val alphaMinValue = 0
        val alphaValue = colorSetting.getAlpha()
        val alphaWidth = size + 18 * scale

        val alphaDiff = min(alphaWidth.toDouble(), max(0.0, mouseX - (getX() - 4 * scale).toDouble()))

        val alphaValueSize = (alphaWidth - 12 * scale) * (alphaValue - alphaMinValue) / (alphaMaxValue - alphaMinValue)

        alphaAnimation.setAnimation(alphaValueSize, 20.0)

        if (alphaDragging) {
            if (colorSetting.isShowAlpha()) {
                colorSetting.setAlpha(
                    if (alphaDiff == 0.0) 0
                    else MathUtils.roundToPlace(
                        (alphaDiff / alphaWidth) * (alphaMaxValue - alphaMinValue) + alphaMinValue,
                        2
                    ).toInt()
                )
            } else {
                colorSetting.setAlpha(255)
            }
        }

        nvgInstance.drawHSBBox(getX(), getY(), size, size, 6f * scale, Color.getHSBColor(colorSetting.getHue(), 1f, 1f))
        nvgInstance.drawRoundedImage(
            ResourceLocation("shindo/hue.png"),
            getX() + 106 * scale,
            getY(),
            12 * scale,
            size,
            3 * scale
        )
        nvgInstance.drawArc(
            getX() + 112 * scale,
            getY() + hueAnimation.getValue() + 6 * scale,
            3 * scale,
            0f,
            360f,
            1.2f * scale,
            Color.WHITE
        )
        nvgInstance.drawArc(
            getX() + saturationAnimation.getValue() + 6 * scale,
            getY() + size - brightnessAnimation.getValue() - 6 * scale,
            3 * scale,
            0f,
            360f,
            1.2f * scale,
            Color.WHITE
        )

        if (colorSetting.isShowAlpha()) {
            nvgInstance.drawRoundedImage(
                ResourceLocation("shindo/alpha.png"),
                getX(),
                getY() + 106 * scale,
                size + 18 * scale,
                12 * scale,
                3 * scale
            )
            nvgInstance.drawAlphaBar(
                getX(),
                getY() + 106 * scale,
                alphaWidth,
                12 * scale,
                3 * scale,
                Color.getHSBColor(colorSetting.getHue(), 1f, 1f)
            )
            nvgInstance.drawArc(
                getX() + alphaAnimation.getValue() + 6 * scale,
                getY() + 112 * scale,
                3 * scale,
                0f,
                360f,
                1.2f * scale,
                Color.WHITE
            )
        }

        nvgInstance.restore()

        nvgInstance.drawRoundedRect(getX() + 106 * scale, getY(), 16 * scale, 16 * scale, 4f, colorSetting.getColor())

        super.draw(mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (isOpen) {
            val size = 100 * scale
            val alphaWidth = size + 18 * scale
            val addY = 26 * scale

            if (mouseButton == 0) {
                if (MouseUtils.isInside(mouseX, mouseY, getX() + 106 * scale, getY() + addY, 12 * scale, size)) {
                    hueDragging = true
                }

                if (MouseUtils.isInside(mouseX, mouseY, getX(), getY() + addY, size, size)) {
                    sbDragging = true
                }

                if (MouseUtils.isInside(
                        mouseX,
                        mouseY,
                        getX(),
                        getY() + 106 * scale + addY,
                        alphaWidth,
                        12 * scale
                    ) && colorSetting.isShowAlpha()
                ) {
                    alphaDragging = true
                }
            }
        }

        if (mouseButton == 0 && isInsideOpen(mouseX, mouseY)) {
            isOpen = !isOpen
        }

        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {
        hueDragging = false
        sbDragging = false
        alphaDragging = false
        super.mouseReleased(mouseX, mouseY, mouseButton)
    }

    fun isShowAlpha(): Boolean = colorSetting.isShowAlpha()

    fun isInsideOpen(mouseX: Int, mouseY: Int): Boolean =
        MouseUtils.isInside(mouseX, mouseY, getX() + 106 * scale, getY(), 16 * scale, 16 * scale)
}